### PR TITLE
Handle BoringUtils in Chisel, rather than in FIRRTL compiler

### PR DESCRIPTION
**Pull request by azidar**
This is a more complete implementation of #3167 , but rather than delaying closing of a module, adds a special area where ports and connections can be added to a module, post-closing. This second phase is closed when calling `.toDefinition` or reflecting on the ports (e.g. with `DataMirror`). 

A longer term plan is to fix some internals (e.g. when closing happens) so that we can unify these phases more gracefully.

For handling the D/I case, we can consider https://github.com/chipsalliance/chisel/pull/3173 as one approach.

### Contributor Checklist

- [x] Did you add Scaladoc to every public function/method?
- [x] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [ ] Did you add appropriate documentation in `docs/src`?
- [x] Did you request a desired merge strategy?
- [x] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement

- Feature (or new API)

#### Desired Merge Strategy

- Squash: The PR will be squashed and merged (choose this if you have no preference.)

#### Release Notes

Added a new BoringUtils.bore that is performed during Chisel elaboration, not via Annotations+CIRCT. Punched ports are accessible to the user via DataMirror. However, using these reflection APIs or calling .toDefinition will fully close a module, to ensure that subsequent boring fails (and thus getting all ports is never stale).

Added a `skipPrefix` to enable ignoring the last prefix value in the prefix name stack.

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels? (Select the most appropriate one based on the "Type of Improvement")
- [ ] Did you mark the proper milestone (Bug fix: `3.5.x` or `3.6.x` depending on impact, API modification or big change: `5.0.0`)?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you do one of the following when ready to merge:
  - [ ] Squash: You/ the contributor `Enable auto-merge (squash)`, clean up the commit message, and label with `Please Merge`.
  - [ ] Merge: Ensure that contributor has cleaned up their commit history, then merge with `Create a merge commit`.

### CODE REVIEW

File: core/src/main/scala/chisel3/IO.scala [hunk 1]
```diff
@@ -44,3 +44,42 @@ object IO {
     iodefClone
   }
 }
+
+private[chisel3] object SecretIO {
+
+  /** Constructs a secret port for the current Module
```
Comment (mwachs5) [inline core/src/main/scala/chisel3/IO.scala:50]: is a SecretIO visible in DataMirror getPorts and friends? if not why not?

File: core/src/main/scala/chisel3/internal/Binding.scala [hunk 2]
```diff
@@ -89,6 +89,10 @@ sealed trait ConditionalDeclarable extends TopBinding {
 @deprecated(deprecatedPublicAPIMsg, "Chisel 3.6")
 case class PortBinding(enclosure: BaseModule) extends ConstrainedBinding
 
+// Added to handle BoringUtils in Chisel
+@deprecated(deprecatedPublicAPIMsg, "Chisel 3.6")
```
Comment (mwachs5) [inline core/src/main/scala/chisel3/internal/Binding.scala:93]: why not make it private right away?

File: src/test/scala/chiselTests/BoringUtilsSpec.scala [hunk 3]
```diff
@@ -125,4 +125,86 @@ class BoringUtilsSpec extends ChiselFlatSpec with ChiselRunners {
     runTester(new InternalBoreTester, annotations = TesterDriver.verilatorOnly) should be(true)
   }
 
+  it should "work using new API" in {
```
Comment (mwachs5) [inline src/test/scala/chiselTests/BoringUtilsSpec.scala:128]: havent read the test in detail, but can you show boring up into a parent and back into a sibling...? or how would that use case be handled?

File: core/src/main/scala/chisel3/IO.scala (hunk 1 above)
Comment (azidar) [reply core/src/main/scala/chisel3/IO.scala:50]: It isn't, and that's to avoid the issue of mutability of a definition. To expand further, if I have code like:
```
val child = Module(new Child)
...
functionFoo(DataMirror.getPorts(child))
...
val myWire = BoringUtils.bore(child.wire)
...
functionBar(DataMirror.getPorts(child))
```

`functionFoo` and `functionBar` would get different ports, which would be difficult to debug because the order of execution matters. This is in general why Chisel has a module "close", so that once I have a reference to `child`, all its Chisel-y things are known.

Another reason supporting not making these present in DataMirror APIs is that all existing uses of boring utils would never show up in a DataMirror getPorts because they were added by CIRCT, so this is backwards compatible behavior.

File: src/test/scala/chiselTests/BoringUtilsSpec.scala (hunk 3 above)
Comment (azidar) [reply src/test/scala/chiselTests/BoringUtilsSpec.scala:128]: The later test shows this is illegal - for right now I'm only letting a user bore downwards.

Comment (azidar) [reply src/test/scala/chiselTests/BoringUtilsSpec.scala:128]: Do I need to support that?

File: core/src/main/scala/chisel3/IO.scala [hunk 4]
```diff
@@ -44,3 +44,42 @@ object IO {
...
+      else
+        try {
+          data.cloneTypeFull
+        } catch {
+          // For now this is going to be just a deprecation so we don't suddenly break everyone's code
+          case e: AutoClonetypeException =>
+            Builder.deprecated(e.getMessage, Some(s"${data.getClass}"))
+            data
+        }
+    module.bindSecretIo(iodefClone)
```
Comment (jackkoenig) [inline core/src/main/scala/chisel3/IO.scala:82]: I know this is from original IO, but I think this chunk of code is old compatibility mode stuff that is no longer necessary.

File: core/src/main/scala/chisel3/Module.scala [hunk 5]
```diff
@@ -259,6 +260,7 @@ package experimental {
   /** Abstract base class for Modules, an instantiable organizational unit for RTL.
     */
   // TODO: seal this?
+  @nowarn("msg=class Port") // delete when Port becomes private
```
Comment (jackkoenig) [inline core/src/main/scala/chisel3/Module.scala:263]: We should probably just promote this warning suppression to build.sbt `-Wconf` level, or go ahead and make the class package private and remove the deprecation.

File: src/test/scala/chiselTests/BoringUtilsSpec.scala (hunk 3 above)
Comment (seldridge) [reply src/test/scala/chiselTests/BoringUtilsSpec.scala:128]: It would be good to allow it to enable the existing Grand Central View API.

This has taps that reach into the parent. Anything in the parent's visible `Data` can be captured. An example use is approximately:

```scala
class Bar extends RawModule {
  val a = Wire(Bool())
}

class Foo extends RawModule {
  val bar = Module(new Bar)
  View {
    val b = Tap(bar.a)
  }
}
```

The result of this is something like the following FIRRTL:

```
module Bar:
  output a_probe: Probe<UInt<1>>
  wire a: UInt<1>

  define(a_probe, probe(a))

module View_companion:
  input a: UInt<1>

  node b = a

module Foo:

  inst bar of Bar
  inst view of View_companion

  view.a <= read(bar.a_probe)
```

File: core/src/main/scala/chisel3/IO.scala (hunk 1 above)
Comment (mwachs5) [reply core/src/main/scala/chisel3/IO.scala:50]: sure, but is this a silent failure of the DataMirror APIs that they would return something different than the user might expect?

File: src/main/scala/chisel3/util/experimental/BoringUtils.scala [hunk 6]
```diff
@@ -193,4 +204,54 @@ object BoringUtils {
...
+            Builder.error(s"Can only bore into modules that have finished building, but $source is in an open module")
+            module = thisModule // exit early
+          } else {
+
+            /** Create a port. */
+            val bore = SecretIO(chiselTypeOf(source), module)
+            module.addSecretIo(bore)
+            module.asInstanceOf[RawModule].secretConnection(bore, rhs)
+
+            rhs = bore
```
Comment (mwachs5) [inline src/main/scala/chisel3/util/experimental/BoringUtils.scala:239]: stupid question -- is this actually doing a connection through the bored IOs? where does that happen?

In other words, its not ust a secret IO that we need, its also a secret connection

File: src/main/scala/chisel3/util/experimental/BoringUtils.scala [hunk 7]
```diff
@@ -193,4 +204,54 @@ object BoringUtils {
...
+
+        if (source.topBinding.location.get == thisModule) {
+
+          /** If no boring actually happened */
+          source
+        } else {
+
+          /** Creating an intermediate wire so secret stuff never escapes */
+          val bore = Wire(chiselTypeOf(source))
+          thisModule.asInstanceOf[RawModule].secretConnection(bore, rhs)
```
Comment (mwachs5) [inline src/main/scala/chisel3/util/experimental/BoringUtils.scala:252]: oh i see here is the secret connection

File: src/test/scala/chiselTests/BoringUtilsSpec.scala (hunk 3 above)
Comment (mwachs5) [reply src/test/scala/chiselTests/BoringUtilsSpec.scala:128]: right, we don't need a "secret IO" for that use case since the modules in question (the View and the Parent) are still open, but the Bore API should support at least one level of up-boring (TBD whether the bore API is supposed to work arbitrarily for some sort of 
"i promise this is accessible from some heirarchy if you keep going up both mine and the boring-top-point you'll find something in common")

File: core/src/main/scala/chisel3/IO.scala (hunk 1 above)
Comment (azidar) [reply core/src/main/scala/chisel3/IO.scala:50]: I can update their documentation to report that they return all non-secret ports?

File: core/src/main/scala/chisel3/Module.scala (hunk 5 above)
Comment (azidar) [reply core/src/main/scala/chisel3/Module.scala:266]: Do you want me to do it in this PR?

File: src/test/scala/chiselTests/BoringUtilsSpec.scala [hunk 8]
```diff
@@ -125,4 +126,145 @@ class BoringUtilsSpec extends ChiselFlatSpec with ChiselRunners {
...
+      "input q_bore : UInt<1>",
+      "q <= q_bore_1", // Do normal connection before secret ones
+      "q_bore_1 <= q_bore",
+      "module Foo :",
+      "input a : UInt<1>",
+      "bar.q_bore <= a"
+    )
+  }
+
+  it should "not affect DataMirror.getPorts, as they are secret (shhh)" in {
```
Comment (mwachs5) [inline src/test/scala/chiselTests/BoringUtilsSpec.scala:221]: this seems a little inconsistent that I am allowed to copy bored ports with toDefinition but I can't reflect on them even before .toDefinitioning? How am I supposed to interact with bored ports on a definition?

File: core/src/main/scala/chisel3/IO.scala (hunk 4 above)
Comment (azidar) [reply core/src/main/scala/chisel3/IO.scala:82]: I think i'll leave it for a future cleanup PR.

Comment (azidar) [reply core/src/main/scala/chisel3/IO.scala:82]: Nvm, just inlined it.

File: core/src/main/scala/chisel3/IO.scala (hunk 1 above)
Comment (azidar) [reply core/src/main/scala/chisel3/IO.scala:50]: Another way we could do this is if `DataMirror.getPorts` calls `.toDefinition` to super-close a module prior to returning all the ports? That way you still get all the ports, but you never get a partial list (e.g. if someone adds ports later, it will error)?

File: core/src/main/scala/chisel3/reflect/DataMirror.scala [hunk 9]
```diff
@@ -346,4 +349,23 @@ object DataMirror {
...
   }
+
+  // Function to find the least common ancestor of two nodes
+  private[chisel3] def leastCommonAncestorModule(left: HasId, right: HasId): Option[BaseModule] = {
+    val leftPathSet = modulePath(left, None).toSet
+    val rightPath = modulePath(right, None)
+    rightPath.collectFirst { case p if leftPathSet.contains(p) => p }
+  }
+  // Function to path upwards, stopping if reaching including
+  private[chisel3] def modulePath(h: HasId, until: Option[BaseModule]): Seq[BaseModule] = {
```
Comment (seldridge) [inline core/src/main/scala/chisel3/reflect/DataMirror.scala:360]: Mega-nit: could you re-order `def modulePath` to occur before it's usage in the LCA method?

File: src/main/scala/chisel3/util/experimental/BoringUtils.scala [hunk 10]
```diff
@@ -193,4 +204,88 @@ object BoringUtils {
...
+    }
+
+    val lca = DataMirror.leastCommonAncestorModule(source, thisModule)
+    if (lca.isEmpty) {
+      Builder.error(s"Cannot bore from $source to ${thisModule.name}, as they do not share a least common ancestor")
+      return Wire(purePortType) //return empty wire to continue elaboration to collect other errors
+    }
+
+    val upPath = DataMirror.modulePath(source, lca)
+    val downPath = DataMirror.modulePath(thisModule, lca)
```
Comment (seldridge) [inline src/main/scala/chisel3/util/experimental/BoringUtils.scala:244]: This organization is a bit odd. The LCA method is used to find the LCA using `modulePath`. Then `modulePath` is re-used to find the path. Would a single method like: `findLCAPaths(foo, bar): (Seq[BaseModule], Seq[BaseModule]) = ???` work better?

File: core/src/main/scala/chisel3/reflect/DataMirror.scala [hunk 11]
```diff
@@ -346,4 +349,23 @@ object DataMirror {
...
+    val me = h match {
+      case m: BaseModule => Seq(m)
+      case d: Data       => d.topBinding.location.toSeq
+      case m: MemBase[_] => m._parent.toSeq
+    }
+    if (me == until.toSeq) Nil
+    else {
+      me ++ me.flatMap(x => x._parent.toSeq.flatMap(p => modulePath(p, until)))
+    }
+  }
```
Comment (seldridge) [inline core/src/main/scala/chisel3/reflect/DataMirror.scala:370]: This method is a bit weird with it's use of `Option` and casts to `Seq` only to `flatMap`. Also, this is using recursion, but is not tail recursive. This would likely be better if rewritten to use iteration.

File: src/main/scala/chisel3/util/experimental/BoringUtils.scala [hunk 12]
```diff
@@ -193,4 +204,88 @@ object BoringUtils {
...
+
+  /** Access a source [[Data]] that may or may not be in the current module.  If
+    * this is in a child module, then create ports to allow access the
+    * requrested source.
+    */
+  def bore(source: Data)(implicit si: SourceInfo): Data = {
+    import reflect.DataMirror
+    def parent(d: Data): experimental.BaseModule = d.topBinding.location.get
+    def purePortType = if (DataMirror.hasOuterFlip(source)) Flipped(chiselTypeOf(source)) else chiselTypeOf(source)
+    def emptyWireOnError = Wire(purePortType) //return empty wire to continue elaboration to collect other errors
```
Comment (seldridge) [inline src/main/scala/chisel3/util/experimental/BoringUtils.scala:216]: Can this return `DontCare` or is that problematic due to that not working with zero-width?

File: src/main/scala/chisel3/util/experimental/BoringUtils.scala [hunk 13]
```diff
@@ -193,4 +204,88 @@ object BoringUtils {
...
+        return emptyWireOnError
+      case Some(internal.CrossModuleBinding) =>
+        Builder.error(
+          s"Cannot bore across a Definition/Instance boundary:${thisModule._errorContext} cannot access ${source}"
+        )
+        return emptyWireOnError
+      case _ => // Actually bore
+    }
+    if (parent(source) == thisModule) {
+
```
Comment (seldridge) [inline src/main/scala/chisel3/util/experimental/BoringUtils.scala:232]: Nit: weird spacing which scalaFmt would clean up.

File: src/main/scala/chisel3/util/experimental/BoringUtils.scala [hunk 14]
```diff
@@ -193,4 +204,88 @@ object BoringUtils {
...
+    }
+    val sink = if (downPath.isEmpty) { lcaSource }
+    else {
+      downPath.reverse.foldLeft(lcaSource) {
+        case (rhs, module) =>
+          chisel3.experimental.skipPrefix { // so `lcaSource` isn't in the name of the secret port
+            if (module.toDefinitionCalled) {
+              Builder.error(
+                s"Can only bore into modules that have not been converted to a Definition: ${module._errorContext}"
+              )
```
Comment (seldridge) [inline src/main/scala/chisel3/util/experimental/BoringUtils.scala:273]: Improving this to track what line did the `toDefinition` would help with debugability. That would make a nice separate PR!

File: src/main/scala/chisel3/util/experimental/BoringUtils.scala [hunk 15]
```diff
@@ -193,4 +204,88 @@ object BoringUtils {
...
+            }
+
+            /** create a port. */
+            val bore = SecretIO(Flipped(purePortType), module)
+            module.addSecretIo(bore)
+            parent(rhs).asInstanceOf[RawModule].secretConnection(bore, rhs)
+            bore
+          }
+      }
+    }
```
Comment (seldridge) [inline src/main/scala/chisel3/util/experimental/BoringUtils.scala:284]: Almost all of these two code paths is the same. Can they be combined?

File: src/main/scala/chisel3/util/experimental/BoringUtils.scala [hunk 16]
```diff
@@ -193,4 +204,88 @@ object BoringUtils {
...
+
+    /** Creating an intermediate wire so secret stuff never escapes */
+    val bore = Wire(purePortType)
+    thisModule.asInstanceOf[RawModule].secretConnection(bore, sink)
```
Comment (seldridge) [inline src/main/scala/chisel3/util/experimental/BoringUtils.scala:288]: What is this doing?

File: src/test/scala/chiselTests/BoringUtilsSpec.scala [hunk 17]
```diff
@@ -125,4 +126,145 @@ class BoringUtilsSpec extends ChiselFlatSpec with ChiselRunners {
...
+      a := BoringUtils.bore(bar.baz.a_wire)
+      b := BoringUtils.bore(bar.b_wire)
+      c := BoringUtils.bore(c_wire)
+    }
+    //println(circt.stage.ChiselStage.emitCHIRRTL(new Foo, args))
+    (circt.stage.ChiselStage
+      .emitCHIRRTL(new Foo)
+      .split('\n')
+      .map(_.takeWhile(_ != '@'))
+      .map(_.trim) should contain).inOrder(
```
Comment (seldridge) [inline src/test/scala/chiselTests/BoringUtilsSpec.scala:159]: This is getting thrown around so much. It would make a nice utility.

File: src/test/scala/chiselTests/BoringUtilsSpec.scala [hunk 18]
```diff
@@ -125,4 +126,145 @@ class BoringUtilsSpec extends ChiselFlatSpec with ChiselRunners {
...
+      "b_bore <= bar.b_bore"
+    )
+  }
+
+  it should "not work over a Definition/Instance boundary" in {
+    import chisel3.experimental.hierarchy._
+    @instantiable
+    class Bar extends RawModule {
+      @public val a_wire = WireInit(UInt(1.W), DontCare)
+      dontTouch(a_wire)
```
Comment (seldridge) [inline src/test/scala/chiselTests/BoringUtilsSpec.scala:181]: Nit: dontTouch is unnecessary here.

File: src/test/scala/chiselTests/BoringUtilsSpec.scala [hunk 19]
```diff
@@ -125,4 +126,145 @@ class BoringUtilsSpec extends ChiselFlatSpec with ChiselRunners {
...
+    }
+    class Foo extends RawModule {
+      val a = IO(Output(UInt()))
```
Comment (seldridge) [inline src/test/scala/chiselTests/BoringUtilsSpec.scala:184]: Nit: `val a` and the connection to it are not necessary.

File: src/test/scala/chiselTests/BoringUtilsSpec.scala [hunk 20]
```diff
@@ -125,4 +126,145 @@ class BoringUtilsSpec extends ChiselFlatSpec with ChiselRunners {
...
+      val bar = Module(new Bar)
+      bar.toDefinition
+      a := BoringUtils.bore(bar.a_wire)
+    }
+
+    val e = intercept[Exception] {
+      circt.stage.ChiselStage.emitCHIRRTL(new Foo, args)
+    }
+    e.getMessage should include("Can only bore into modules that have not been converted to a Definition")
+  }
```
Comment (seldridge) [inline src/test/scala/chiselTests/BoringUtilsSpec.scala:269]: Nice error test case. This could also be cleaned up by removing the dontTouch and port `a`.

**Review by seldridge — COMMENTED** [review]
This is shaping up great!

Minor comments throughout. I do think that the second-class nature of `DataMirror` relative to `toDefinition` may not be a great API. I was always okay with `DataMirror` and `toDefinition` having the "effect" of making the ports visible (which is roughly the same as causing the module to close).

File: core/src/main/scala/chisel3/Module.scala [hunk 21]
```diff
@@ -312,6 +317,8 @@ package experimental {
     /** Internal check if a Module is closed */
     private[chisel3] def isClosed = _closed
 
+    private[chisel3] var toDefinitionCalled = false
+
```
[the remark this answers is not in the document]
Comment (azidar) [reply core/src/main/scala/chisel3/Module.scala:321]: The reason I'm making the distinction is that the `_closed` check means that the normal phase is done, so creating `Wires` etc in a module should be illegal. It has implications for when the `component` is generated, as well as other kinds of checks internally.

File: src/main/scala/chisel3/util/experimental/BoringUtils.scala (hunk 12 above)
Comment (azidar) [reply src/main/scala/chisel3/util/experimental/BoringUtils.scala:216]: Hmm. This is only returned on the case when we are definitely throwing an error, but probably `DontCare` will work, I'll try it.

File: src/test/scala/chiselTests/BoringUtilsSpec.scala (hunk 8 above)
Comment (azidar) [reply src/test/scala/chiselTests/BoringUtilsSpec.scala:221]: Agreed, I'm fixing it to work the same way, that reflecting "fully closes the modules"

File: src/main/scala/chisel3/util/experimental/BoringUtils.scala (hunk 14 above)
Comment (azidar) [reply src/main/scala/chisel3/util/experimental/BoringUtils.scala:273]: I'll add it to this PR, it's not too hard.

File: core/src/main/scala/chisel3/IO.scala (hunk 1 above)
Comment (azidar) [reply core/src/main/scala/chisel3/IO.scala]: Added clear error messages, good discussion, thanks!

File: src/test/scala/chiselTests/ChiselSpec.scala [hunk 22]
```diff
@@ -360,3 +360,22 @@ trait Utils {
...
+    val lines = output.split("\n").toList
+    val unmatched = matchList.flatMap { m =>
+      if (matches(lines, m).nonEmpty) None else Some(m)
+    }.map(x => s"  > $x was unmatched")
+    val unomitted = omitList.flatMap { o => omits(lines, o) }.map {
+      case (o, l) => s"  > $o was not omitted in ($l)"
+    }
+    val results = unmatched ++ unomitted
+    assert(results.isEmpty, results.mkString("\n"))
+  }
```
Comment (seldridge) [inline src/test/scala/chiselTests/ChiselSpec.scala:380]: Mega-nit: isn't this `matchesAndOmits`?

Non-blocking feedback: the multiple parameter lists can be hard to grok, but are necessary to have multiple variadic arguments. Using Arrays and named arguments when using this method may make it more obvious.

Alternatively, just having a FIRRTL trimming/canonicalizing utility and letting users do their own ScalaTest matching is more what I was thinking here.

File: src/test/scala/chiselTests/naming/PrefixSpec.scala [hunk 23]
```diff
@@ -67,6 +67,38 @@ class PrefixSpec extends ChiselPropSpec with Utils {
...
+          "x1_wire1",
+          "x1",
+          "wire1",
+          "x2",
+          "wire1_1",
+          "wire2"
+        )
+      )
+    }
+  }
```
Comment (seldridge) [inline src/test/scala/chiselTests/naming/PrefixSpec.scala:100]: An `emitCHIRRTL` test would be cleaner, IMO.

**Review by seldridge — APPROVED** [review]
LGTM

File: core/src/main/scala/chisel3/Module.scala [hunk 24]
```diff
@@ -537,6 +557,35 @@ package experimental {
...
+    // data must be a fresh Chisel type
+    private[chisel3] def createSecretIO(data: => Data)(implicit sourceInfo: SourceInfo): Data = {
+      val iodef = data
+      // forcename
+      internal.requireIsChiselType(iodef, "io type")
+
+      if (toDefinitionCalled.nonEmpty)
+        Builder.error("can only create secret ports when a module has not been used as a Definition")
+      if (modulePortsAskedFor.nonEmpty)
+        Builder.error("can only create secret ports when a module has not been asked for all its IO")
```
[the remark this answers is not in the document]
Comment (azidar) [reply core/src/main/scala/chisel3/Module.scala:570]: My mistake, this is an internal error, i'll switch to require.

**Comment by sequencer** [comment]
I wonder if it’s possible to use RefOp + XMR to replace the BoringUtils?

**Comment by seldridge** [comment]
> I wonder if it’s possible to use RefOp + XMR to replace the BoringUtils?

We want both! This `bore` API is going to do normal port boring. We'll add a second API, `tap`, that will use the same algorithm, but do ref type ports on the upwards connection once https://github.com/chipsalliance/chisel/pull/3088 lands.

I.e., this is the exact same behavior as FIRRTL Dialect's `WiringProblem`. The awesome work Adam is doing here enables us to do all this work in Chisel.

**Comment by sequencer** [comment]
Sounds great!
